File: An2/MPP/mpp-proiect-java-DeepSmeag/persistenceModule/src/main/java/contests/persistence/CategorieVarstaRepo.java
package contests.persistence;

import contests.model.CategorieVarsta;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import contests.persistence.interfaces.CategorieVarstaRepoInterface;
import contests.persistence.utils.DBConnection;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

public class CategorieVarstaRepo implements CategorieVarstaRepoInterface {
    private static final Logger logger = LogManager.getLogger();

    @Override
    public List<CategorieVarsta> getAll() {
        try (Connection connection = DBConnection.getNewConnection();
             PreparedStatement ps = connection.prepareStatement("SELECT * FROM \"CategoriiVarsta\"")) {
            ResultSet resultSet = ps.executeQuery();
            ArrayList<CategorieVarsta> categories = new ArrayList<>();
            while (resultSet.next()) {
                Integer id = resultSet.getInt("id");
                int ageMin = resultSet.getInt("min");
                int ageMax = resultSet.getInt("max");
                CategorieVarsta category = new CategorieVarsta(id, ageMin, ageMax);
                logger.info("Successfully retrieved category with id " + id + " from database.");
                categories.add(category);
            }
            return categories;
        } catch (SQLException e) {
            logger.error("Error while retrieving categories from database.");
            return null;
        }
    }

    @Override
    public CategorieVarsta getOne(Integer id) {
        try (Connection connection = DBConnection.getNewConnection();
             PreparedStatement ps = connection.prepareStatement("SELECT * FROM \"CategoriiVarsta\" WHERE id=" + String.valueOf(id) + ";")) {
//            ps.setInt(0, id);
            ResultSet resultSet = ps.executeQuery();

            resultSet.next();
            int ageMin = resultSet.getInt("min");
            int ageMax = resultSet.getInt("max");
            CategorieVarsta category = new CategorieVarsta(id, ageMin, ageMax);
//                Log.d("tag", "restul");


            return category;
        } catch (SQLException e) {
//            throw new RuntimeException(e);
//            Log.E("ErRROR","ERR");
            return null;
        }
    }

    @Override
    public boolean saveEntity(CategorieVarsta entity) {
        try (Connection connection = DBConnection.getNewConnection();
             PreparedStatement ps = connection.prepareStatement("INSERT INTO \"CategoriiVarsta\" (\"min\", \"max\") VALUES(?,?)")) {
            ps.setInt(0, entity.getVarstaMinima());
            ps.setInt(1, entity.getVarstaMaxima());
            ps.executeUpdate();
//                Log.d("tag", "restul");
            return true;
        } catch (SQLException e) {
//            throw new RuntimeException(e);
//            Log.E("ErRROR","ERR");
            return false;
        }
    }

    @Override
    public boolean deleteEntity(Integer id) {
        try (Connection connection = DBConnection.getNewConnection();
             PreparedStatement ps = connection.prepareStatement("DELETE FROM \"CategoriiVarsta\" WHERE id=?")) {
            ps.setInt(0, id);
            ps.executeUpdate();
//                Log.d("tag", "restul");
            return true;
        } catch (SQLException e) {
//            throw new RuntimeException(e);
//            Log.E("ErRROR","ERR");
            return false;
        }
    }

    @Override
    public boolean updateEntity(CategorieVarsta entity) {
        try (Connection connection = DBConnection.getNewConnection();
             PreparedStatement ps = connection.prepareStatement("UPDATE \"CategoriiVarsta\" SET (\"min\"=?, \"max\"=?) WHERE id=?")) {
            ps.setInt(0, entity.getVarstaMinima());
            ps.setInt(1, entity.getVarstaMaxima());
            ps.setInt(2, entity.getId());
            ps.executeUpdate();
//                Log.d("tag", "restul");
            return true;
        } catch (SQLException e) {
//            throw new RuntimeException(e);
//            Log.E("ErRROR","ERR");
            return false;
        }
    }

    @Override
    public CategorieVarsta getCategorieVarstaByAgeGroup(int ageMin, int ageMax) {
        try (Connection connection = DBConnection.getNewConnection();
             PreparedStatement ps = connection.prepareStatement("SELECT * FROM \"CategoriiVarsta\" WHERE \"min\"=? AND \"max\"=?")) {
            ps.setInt(0, ageMin);
            ps.setInt(1, ageMax);
            ResultSet resultSet = ps.executeQuery();

            resultSet.next();
            int id = resultSet.getInt("id");
            CategorieVarsta category = new CategorieVarsta(id, ageMin, ageMax);
//                Log.d("tag", "restul");


            return category;
        } catch (SQLException e) {
//            throw new RuntimeException(e);
//            Log.E("ErRROR","ERR");
            return null;
        }
    }
}

File: An2/MPP/mpp-proiect-csharp-DeepSmeag/persistence/CategorieVarstaRepo.cs
using mpp_proiect_csharp_DeepSmeag.persistence.interfaces;
using mpp_proiect_csharp_DeepSmeag.persistence.utils;
using mpp_proiect_csharp_DeepSmeag.models;
using Npgsql;

namespace mpp_proiect_csharp_DeepSmeag.persistence
{
    public class CategorieVarstaRepo : ICategorieVarstaRepoInterface
    {
        private static readonly log4net.ILog log = log4net.LogManager.GetLogger(System.Reflection.MethodBase.GetCurrentMethod().DeclaringType);
        public bool deleteEntity(int entityId)
        {
            try
            {
                using var con = DBUtils.GetNewConnection();
                con.Open();
                var sqlString = "DELETE FROM \"CategoriiVarsta\" WHERE id=@id";
                using var sqlCommand = new NpgsqlCommand(sqlString, con);
                sqlCommand.Parameters.AddWithValue("id", entityId);
                sqlCommand.Prepare();
                sqlCommand.ExecuteNonQuery();
                con.Close();
                return true;


            }
            catch (Exception)
            {
                return false;
            }
        }

        public List<CategorieVarsta>? getAll()
        {
            try
            {
                using var con = DBUtils.GetNewConnection();
                con.Open();
                var sqlString = "SELECT * FROM \"CategoriiVarsta\"";
                using var sqlCommand = new NpgsqlCommand(sqlString, con);
                using NpgsqlDataReader rdr = sqlCommand.ExecuteReader();
                List<CategorieVarsta> categoriiVarsta = new List<CategorieVarsta>();
                while (rdr.Read())
                {
                    int id = rdr.GetInt32(0);
                    int ageMin = rdr.GetInt32(1);
                    int ageMax = rdr.GetInt32(2);
                    CategorieVarsta categorieVarsta = new CategorieVarsta(id, ageMin, ageMax);
                    categoriiVarsta.Add(categorieVarsta);
                    log.Info("Succes la adaugare");
                }
                return categoriiVarsta;
            }
            catch (Exception) {
                log.Error("Eroare la retrieval");
                return null; 
            }
        }

        public CategorieVarsta? getCategorieVarstaByAgeGroup(int minAge, int maxAge)
        {
            try
            {
                using var con = DBUtils.GetNewConnection();
                con.Open();
                var sqlString = "SELECT * FROM \"CategoriiVarsta\" WHERE \"min\"=@ageMin AND \"max\"=@ageMax";
                using var sqlCommand = new NpgsqlCommand(sqlString, con);
                sqlCommand.Parameters.AddWithValue("ageMin", minAge);
                sqlCommand.Parameters.AddWithValue("ageMax", maxAge);
                sqlCommand.Prepare();
                using NpgsqlDataReader rdr = sqlCommand.ExecuteReader();

                rdr.Read();
                int id = rdr.GetInt32(0);
                CategorieVarsta categorieVarsta = new CategorieVarsta(id, minAge, maxAge);

                return categorieVarsta;
            }
            catch (Exception) { return null; }
        }

        public CategorieVarsta? getOne(int entityId)
        {
            try{
                using var con = DBUtils.GetNewConnection();
                con.Open();
                var sqlString = "SELECT * FROM \"CategoriiVarsta\" WHERE id=@id";
                using var sqlCommand = new NpgsqlCommand(sqlString, con);
                sqlCommand.Parameters.AddWithValue("id", entityId);
                sqlCommand.Prepare();
                using NpgsqlDataReader rdr = sqlCommand.ExecuteReader();

                rdr.Read();
                int id = rdr.GetInt32(0);
                int ageMin = rdr.GetInt32(1);
                int ageMax = rdr.GetInt32(2);
                CategorieVarsta categorieVarsta = new CategorieVarsta(id, ageMin, ageMax);

                return categorieVarsta;
            }
            catch (Exception) { return null; }
        }

        public bool saveEntity(CategorieVarsta entityType)
        {
            try {
                using var con = DBUtils.GetNewConnection();
                con.Open();
                var sqlString = "INSERT INTO \"CategoriiVarsta\" (\"min\", \"max\") VALUES (@ageMin, @ageMax)";
                using var sqlCommand = new NpgsqlCommand(sqlString, con);
                sqlCommand.Parameters.AddWithValue("ageMin", entityType.VarstaMinima);
                sqlCommand.Parameters.AddWithValue("ageMax", entityType.VarstaMaxima);
                sqlCommand.Prepare();
                sqlCommand.ExecuteNonQuery();
                con.Close();
                return true;
            }
            catch (Exception) { return false; }
        }

        public bool updateEntity(CategorieVarsta entityType)
        {
            try {
                using var con = DBUtils.GetNewConnection();
                con.Open();
                var sqlString = "UPDATE \"CategoriiVarsta\" SET \"min\"=@ageMin, \"max\"=@ageMax WHERE id=@id";
                using var sqlCommand = new NpgsqlCommand(sqlString, con);
                sqlCommand.Parameters.AddWithValue("id", entityType.Id);
                sqlCommand.Parameters.AddWithValue("ageMin", entityType.VarstaMinima);
                sqlCommand.Parameters.AddWithValue("ageMax", entityType.VarstaMaxima);
                sqlCommand.Prepare();
                sqlCommand.ExecuteNonQuery();
                con.Close();
                return true;
            }
            catch (Exception) { return false; }
        }

        CategorieVarsta? IRepoInterface<CategorieVarsta, int>.getOne(int entityId)
        {
            throw new NotImplementedException();
        }
    }
}
